common/achievement/util.ts

```diff
@@ -1,28 +1,54 @@
-import { ActionID, NotificationContext } from '../notification/types';
+import { ActionID } from '../notification/types';
 import { metricsByAction } from './metrics';
 import { Metric } from './types';
+import { prisma } from '../prisma';
+import { getLogger } from '../logger/logger';
 
-export function getRelationByContext(context: NotificationContext): string {
-    const { appointment, match, subcourse } = context;
-    let relation: string;
-
-    switch (context) {
-        case context.match:
-            relation = `match/${match.id}`;
-            break;
-        case context.subcourse:
-            relation = `subcourse/${subcourse.id}`;
-            break;
-        case context.appointment:
-            relation = `appointment/${appointment.id}`;
-            break;
-        default:
-            relation = '';
+const logger = getLogger('Gamification');
+export function isGamificationFeatureActive(): boolean {
+    const isActive: boolean = JSON.parse(process.env.GAMIFICATION_ACTIVE || 'false');
+
+    if (!isActive) {
+        logger.warn('Gamification is deactivated');
     }
 
-    return relation;
+    return isActive;
+}
+
+export function assureGamificationFeatureActive() {
+    if (!isGamificationFeatureActive()) {
+        throw new Error(`Gamification is deactivated`);
+    }
 }
 
 export function getMetricsByAction<ID extends ActionID>(actionId: ID): Metric[] {
     return metricsByAction.get(actionId) || [];
 }
+
+type RelationTypes = 'match' | 'subcourse';
+
+export function getRelationTypeAndId(relation: string): [relationType: RelationTypes, id: number] {
+    const validRelationTypes = ['match', 'subcourse'];
+    const [relationType, relationId] = relation.split('/');
+    if (!validRelationTypes.includes(relationType)) {
+        throw Error('No valid relation found in relation: ' + relationType);
+    }
+
+    const parsedRelationId = parseInt(relationId, 10);
+    return [relationType as RelationTypes, parsedRelationId];
+}
+
+export async function getRelationContext(relation: string) {
```

Comment:
What do you think about creating a common response object here? Something like
```
{
   match?: Match,
   course?: Subcourse,
}
```

This will make sure that we can extend the return data and we don't have to rely on casting the values.

Reply:
change the getRelationContext function:
```
export async function getRelationContext(relation: string): Promise<RelationContextType> {
    const [type, id] = getRelationTypeAndId(relation);
    const relationContext: RelationContextType = {
        match: type === 'match' ? await prisma.match.findFirst({ where: { id }, select: { id: true, lecture: { select: { start: true, duration: true } } } })[0] : null,
        subcourse: type === 'subcourse' ? await prisma.subcourse.findFirst({ where: { id }, select: { id: true, lecture: { select: { start: true, duration: true } } } })[0] : null,
        actionNames: type !== 'match' && type !== 'subcourse' ? relation.split(',') : null,
    };
    return relationContext;
}
```
We now return the same type no matter what relation is. The receiving parent function makes the decision on how to handle the result.